kernel: Separate UTXO set access from validation functions

sedited opened the pull request:
The current code makes it hard to call functions for checking the consensus-correctness of a block without having a full UTXO set at hand. This makes implementing features such as Utreexo and non-assumevalid swiftsync, where maintaining a full UTXO set defeats their purpose, through the kernel API difficult. Checks like the coinbase subsidy, or block sigops are only available through `ConnectBlock`, which requires a complete coins cache.

Solve this by moving accessing coins and the responsibility of spending them out of `ConnectBlock` and into a new separate function called `SpendBlock`. Pass a block's spent coins through the existing `CBlockUndo` data structure to `ConnectBlock`. 

While the change is largely a refactor, some smaller behaviour changes were unavoidable. These should be highlighted in the commit messages. The last commit contains the most important part of the coins logic move, the commits before attempt to prepare it piecemeal.

This pull request was benchmarked on [benchcoin](https://github.com/bitcoin-dev-tools/benchcoin/pull/162), which showed no performance regression. While discussing this pull request it was noted that pre-fetching and eliminating the need for some additional map lookups might improve performance a bit, but this does not make a measurable difference in the grand scheme of things and was not the motivation for this pull request.

In future changes, ConnectBlock could be made even more self contained, accessing the coins from the internal map could be further optimized by consolidating existence checking and spending, and the function eventually exposed in an external kernel API (as done [here](https://github.com/sedited/bitcoin/commit/09e05493010119fc60d310d3e7ffc388de441f9a#diff-f638e65d9278276b2a6ac42ffbbd26d079c3dba752feed5038a25a1ee565d320R947))

sedited committed:
consensus: Use Coin span in GetP2SHSigOpCount

Move the responsibility of retrieving coins from GetP2SHSigOpCount to
its caller.

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

Define two explicit template specializations for both a span of
references to coins and a span of coins. This allows using it for both
Coin entries referenced from the CCoinsViewCache, and from contiguous
memory, like the vector in CBlockUndo.

sedited committed:
consensus: Use Coin span in GetTransactionSigOpCost

Move the responsibility of retrieving coins from GetTransactionSigOpCost
to its caller.

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

Define two explicit template specializations for both a span of
references to coins and a span of coins. This allows using it for both
Coin entries referenced from the CCoinsViewCache, and from contiguous
memory, like the vector in CBlockUndo.

sedited committed:
validation: Move coin existence and spend check to SpendBlock

Move the remaining UTXO-related operations from ConnectBlock to
SpendBlock. This includes moving the existence check, the UpdateCoins
call, and CBlockUndo generation.

ConnectBlock now takes a pre-populated CBlockUndo as an argument and no
longer accesses or manipulates the UTXO set.

Co-authored-by: w0xlt <94266259+w0xlt@users.noreply.github.com>

sedited committed:
consensus: Use Coin span in CheckTxInputs

Move the responsibility of retrieving coins from CheckTxInputs
to its caller. The missing inputs check will be moved in an upcoming
commit.

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

Define two explicit template specializations for both a span of
references to coins and a span of coins. This allows using it for both
Coin entries referenced from the CCoinsViewCache, and from contiguous
memory, like the vector in CBlockUndo.

sedited committed:
validation: Use vector of outputs instead of CCoinsViewCache in CheckInputScripts

Move the responsibility of retrieving coins from the CCoinsViewCache
in CheckInputScripts to its caller.

Add a helper method in CCoinsViewCache to collect all Coin's outputs
spent by a transaction's inputs.

Callers of CCoinsViewCache are updated to either pre-fetch the spent
outputs, or pass in an empty vector if the precomputed transaction data
has already been initialized with the required outputs.

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

sedited committed:
validation: Add SpendBlock function

Move the BIP30 checks from ConnectBlock to a new SpendBlock method.
This is a move-only change, more content to SpendBlock is added in the
next commits. The goal is to move logic requiring access to the
CCoinsViewCache out of ConnectBlock and to the new SpendBlock method.
SpendBlock will in future handle all UTXO set interactions that
previously took place in ConnectBlock.

Callers of ConnectBlock now also need to call SpendBlock before. This is
enforced in a future commit by adding a CBlockUndo argument that needs
to be passed to both.

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

Co-authored-by: w0xlt <94266259+w0xlt@users.noreply.github.com>

sedited committed:
validation: Move SetBestBlock out of ConnectBlock

This is a part of a series of commits for removing access to the
CCoinsViewCache in consensus verification functions. The goal is to
allow calling verification functions with pre-fetched, or a user-defined
set of coins.

Sjors commented on src/validation.h:
```diff
@@ -709,6 +709,9 @@ class Chainstate
     bool ConnectBlock(const CBlock& block, BlockValidationState& state, CBlockIndex* pindex,
                       CCoinsViewCache& view, bool fJustCheck = false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    bool SpendBlock(const CBlock& block, const CBlockIndex* pindex, const uint256& block_hash,
```
2be3c59151bcb840eb8cbd045a3dada6775ebc9e: this or the final commit would be a good time to document both `SpendBlock` and `ConnectBlock`.

TheCharlatan replied:
Yes, that is the gist of it. Will add a commit at the end to add some documentation in the header.

TheCharlatan replied:
Added a commit.

darosior commented on src/validation.h:
```diff
@@ -706,8 +706,11 @@ class Chainstate
-    bool ConnectBlock(const CBlock& block, BlockValidationState& state, CBlockIndex* pindex,
-                      CCoinsViewCache& view, bool fJustCheck = false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    bool ConnectBlock(const CBlock& block, const uint256& block_hash, BlockValidationState& state, CBlockIndex* pindex,
+                      CBlockUndo& blockundo, bool fJustCheck = false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
I think it would be clearer to take a vector of spent `Coin`s directly? Or even a span?

TheCharlatan replied:
I might be missing something, but it would have to be a span of `CTxUndo`, and you'd have to move `WriteBlockUndo` to `ConnectTip` too, no? I wasn't quite sure about doing that change here though. The implications of separately writing the undo data and raising the `BlockIndex` validity level are not quite clear to me yet. There is always a question where to draw the scope line with these refactors. While it might not be useful to write undo data if we ever get to the alternative kernel API client use cases, there might be other ways to handle that.

darosior replied:
Oh, i was just reading the diff and overlooked `WriteBlockUndo`! Yes that makes sense.

sedited committed:
doc: Add docstrings for ConnectBlock and SpendBlock

w0xlt reviewed (commented):
Concept ACK, as this PR increases the decoupling of components.

nit: The name `SpendBlock` may not be clear (unless this concept is already known and I'm missing the point).
Maybe `BIP30Validation`, something like that, would be clearer.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/32317#pullrequestreview-2785006540

> nit: The name SpendBlock may not be clear (unless this concept is already known and I'm missing the point).
Maybe BIP30Validation, something like that, would be clearer.


It is doing more than BIP-30 validation, so I don't think that would accurate either. I thought the name was pretty clear for what it does, but I'd be open to other suggestions.

stringintech commented on src/validation.cpp:
```diff
@@ -2161,7 +2172,7 @@ ValidationCache::ValidationCache(const size_t script_execution_cache_bytes, cons
  * Non-static (and redeclared) in src/test/txvalidationcache_tests.cpp
  */
 bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
-                       const CCoinsViewCache& inputs, unsigned int flags, bool cacheSigStore,
+                       std::vector<CTxOut>&& spent_outputs, unsigned int flags, bool cacheSigStore,
                        bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
```
Is there a reason we keep passing the `spent_outputs` instead of removing it (to simplify the API) and requiring callers to initialize `txdata` before calling `CheckInputScripts`?

TheCharlatan replied:
My read is that the call to `txdata.Init` is delayed to after checking the script execution cache for performance reasons. Calling `Init` does some hashing, that would not be required if it is already in the cache. Moving the `spent_outputs` vector construction is already a potential performance downside for potential cache hits, though I think that is more acceptable in the grand scheme of things.

stringintech replied:
Oh, it makes sense. Thanks for the insights.

TheCharlatan commented on src/consensus/tx_verify.h:
```diff
@@ -39,11 +40,12 @@ unsigned int GetLegacySigOpCount(const CTransaction& tx);
  *
- * @param[in] mapInputs Map of previous transactions that have outputs we're spending
+ * @param[in] coins Sorted span of Coins containing previous transaction outputs we're spending
  * @return maximum number of sigops required to validate this transaction's inputs
  * @see CTransaction::FetchInputs
  */
-unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& mapInputs);
+template <typename T>
+unsigned int GetP2SHSigOpCount(const CTransaction& tx, const T coins);
```
Yeah, that seems better. I had another version of this that tried to enforce everything through concepts, but thought that didn't really help with keeping things clear, so just dropped all the fancy type hints altogether.

TheCharlatan commented:
Updated 76a8f22c5c5cf48a9c36cc40db9224f0454917d0 -> 16a695fbff4a92eba3eb72ec6aa766e71c52f773 ([spendblock_0](https://github.com/TheCharlatan/bitcoin/tree/spendblock_0) -> [spendblock_1](https://github.com/TheCharlatan/bitcoin/tree/spendblock_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_0..spendblock_1))

* Addressed @stringintech's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2078426205), now using `std::span<T>` as templated argument.

stickies-v commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -124,6 +124,9 @@ static void ValidateCheckInputsForAllFlags(const CTransaction &tx, uint32_t fail
 {
     PrecomputedTransactionData txdata;
 
+    std::vector<CTxOut> spent_outputs{active_coins_tip.GetUnspentOutputs(tx)};
+    txdata.Init(tx, std::move(spent_outputs));
```
In a way, it's a bit silly to add a parameter to the `CheckInputScripts` signature to then just move it into another parameter (`txdata`). An alternative approach could be to let the caller handle this, and `assert txdata.m_spent_outputs_ready`, but that could raise runtime assertion errors instead of catching them at compile time.

stringintech replied:
As @TheCharlatan noted [here](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2083133917), delaying `txdata.Init` could be on purpose for the potential cache hit [here](https://github.com/TheCharlatan/bitcoin/blob/16a695fbff4a92eba3eb72ec6aa766e71c52f773/src/validation.cpp#L2195:L2201). I wonder if there is a cleaner version for achieving the same goal though.

stickies-v reviewed (commented):
Concept ACK

Reducing the coupling between validation functions and the UTXO set makes the validation functions more usable, and more testable. The approach taken here seems sensible, but I'm going to think about alternatives a bit more.

stringintech commented on src/consensus/tx_verify.cpp:
```diff
@@ -14,6 +14,16 @@
 #include <util/check.h>
 #include <util/moneystr.h>
 
+template <typename T>
+static constexpr const Coin& GetCoin(const T& item)
+{
+    if constexpr (std::is_same_v<T, std::reference_wrapper<const Coin>>) {
+        return item.get();
+    } else {
+        return item;
+    }
+}
+
```
It compiles for me as well.

TheCharlatan replied:
Pushed the compiler issue on a separate branch here: https://github.com/TheCharlatan/bitcoin/actions/runs/15016860100/job/42196724477#step:10:198 .

TheCharlatan replied:
Sweet, will apply this!

stringintech commented on src/validation.cpp:
```diff
@@ -2687,7 +2594,12 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+        if (!tx.IsCoinBase()) {
+            nSigOpsCost += GetTransactionSigOpCost(tx, std::span<const Coin>{blockundo.vtxundo[i - 1].vprevout}, flags);
+        } else {
+            std::vector<Coin> coin;
+            nSigOpsCost += GetTransactionSigOpCost(tx, std::span<const Coin>{coin}, flags);
```
Looks like we have an early return in `GetTransactionSigOpCost` for coinbase tx and no use for `coins` span; so maybe we could just pass an empty span here?
```suggestion
            nSigOpsCost += GetTransactionSigOpCost<const Coin>(tx, {}, flags);
```

stringintech commented on src/consensus/tx_verify.cpp:
```diff
@@ -123,62 +135,73 @@ unsigned int GetLegacySigOpCount(const CTransaction& tx)
+template int64_t GetTransactionSigOpCost<const Coin>(
+    const CTransaction& tx, std::span<const Coin> coins, uint32_t flags);
 
-bool Consensus::CheckTxInputs(const CTransaction& tx, TxValidationState& state, const CCoinsViewCache& inputs, int nSpendHeight, CAmount& txfee)
-{
-    // are the actual inputs available?
-    if (!inputs.HaveInputs(tx)) {
-        return state.Invalid(TxValidationResult::TX_MISSING_INPUTS, "bad-txns-inputs-missingorspent",
-                         strprintf("%s: inputs missing/spent", __func__));
-    }
```
Do we need to update the [documentation](https://github.com/TheCharlatan/bitcoin/blob/16a695fbff4a92eba3eb72ec6aa766e71c52f773/src/consensus/tx_verify.h#L24:L31) for `CheckTxInputs` now that the input existence check is removed from it?

stringintech commented on src/validation.cpp:
In commit 5908680, the description mentions: "By returning early in case of a BIP30 validation failure, an additional failure check for an invalid block reward in case of its failure is left out."

To validate my understanding: there are also other checks in `ConnectBlock` before the `!state.IsValid()` check that we may have skipped because of the early return when `SpendBlock` fails - including transaction input validation, script verification, fee range checks, ...; Is that correct?

It makes sense to me not to proceed with additional validation if `SpendBlock` already detects a consensus failure, but I'm curious why the previous code continued checking for other consensus violations even after detecting an early BIP30 failure?

TheCharlatan replied:
> There are also other checks in ConnectBlock before the !state.IsValid() check that we may have skipped because of the early return when SpendBlock fails - including transaction input validation, script verification, fee range checks, ...; Is that correct?

Good catch! Actually I think the commit message description is misleading, if not wrong. The checks you mention are skipped in any case, because we check `if (!state.IsValid()) break;`, but similarly the coinbase check won't override the validation result if it is already invalid. The check is still executed though before returning.

stringintech reviewed (commented):
Concept ACK.
I had a few more comments (sorry for the scattered feedback; didn't have the chance to review everything earlier).

stickies-v commented:
PSA: Bitcoin Core PR Review Club will cover this PR in its next meeting at 2025-05-14 at 17:00 UTC. See https://bitcoincore.reviews/32317 for notes, questions, and instructions on [how to join](https://bitcoincore.reviews/).

TheCharlatan commented on src/validation.h:
```diff
@@ -706,8 +706,36 @@ class Chainstate
+     * undo data, and raises the block validity in the block index.
+     *
+     * @param[in] block       The block to be validated
+     * @param[in] block_hash  Hash of block
```
Yes, that is it. I don't think this will actually matter in the grand scheme. I think I'll drop it again.

TheCharlatan replied:
Dropped it again, but I also changed things slightly to get the block hash from the block index now.

stickies-v commented on src/consensus/tx_verify.cpp:
```diff
@@ -123,62 +133,73 @@ unsigned int GetLegacySigOpCount(const CTransaction& tx)
+    const CTransaction& tx, const std::span<std::reference_wrapper<const Coin>>);
+
+template <typename T>
+int64_t GetTransactionSigOpCost(const CTransaction& tx, const T coins, uint32_t flags)
```
Imo, there are 2 main advantages:
- discoverability: grepping for callsites, template instantiations etc is a lot more cumbersome (and easy to miss) than just inspecting the canonical template definition
- type safety: explicitly narrowing down the actual types that `T` can be makes reviewing the function easier, because you have to think about edge cases less, e.g. will the function still be safe when a new callsite is added

This is more a general style preference, and not particularly important for this PR, so feel free to ignore / can mark as resolved.

TheCharlatan commented:
Thank you for the reviews @stringintech and @stickies-v!

16a695fbff4a92eba3eb72ec6aa766e71c52f773 -> e8ac6d29273c48328f1f77886c3c3f653e6cc58e ([spendblock_1](https://github.com/TheCharlatan/bitcoin/tree/spendblock_1) -> [spendblock_2](https://github.com/TheCharlatan/bitcoin/tree/spendblock_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_1..spendblock_2))

* Addressed @stickies-v's comments [1](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2065965592), [2](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2080154034), [3](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2086983097), [4](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2086984869), [5](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2086978979) improving docstrings for `ConnectBlock` and `SpendBlock`.
* Addressedd @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2066960211), removed block_hash argument from `ConnectBlock` and `SpendBlock` methods.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2075895517), added `CoinRef` concept.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2075811232), removed unneeded helper, use direct initialization to solve msvc compilation error.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2086974181), add assert for `blockundo.vtxundo.size()` to rule out UB.
*Addressed @stringintech's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2087539901), pass in empty span to `GetTransactionSigOpCost`.
* Addressed @stringintech's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2087577718), update documentation for the removal of the inputs availability check in `CheckTxInputs`.
* Addressed @stringintech's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2087608225), removed line in commit message mentioning execution of the coinbase amount check even when another check has failed previously.

TheCharlatan commented on src/coins.cpp:
```diff
@@ -160,6 +160,29 @@ const Coin& CCoinsViewCache::AccessCoin(const COutPoint &outpoint) const {
+{
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(tx.vin.size());
+    for (const auto& txin : tx.vin) {
+        const COutPoint& prevout = txin.prevout;
+        const Coin& coin = AccessCoin(prevout);
+        assert(!coin.IsSpent());
+        spent_outputs.emplace_back(coin.out);
+    }
```
We could also use `AccessCoin`, but I think accessing it directly is a bit more efficient, because we don't allocate another vector. Do you have a better suggestion for the naming there? It is all a bit confusing.

TheCharlatan commented:
Rebased e8ac6d29273c48328f1f77886c3c3f653e6cc58e -> e604a57dccac1c4e6802a3ecc7f4cb4c3229c8ed ([spendblock_2](https://github.com/TheCharlatan/bitcoin/tree/spendblock_2) -> [spendblock_3](https://github.com/TheCharlatan/bitcoin/tree/spendblock_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_2..spendblock_3))

* Fixed conflict with #32575

TheCharlatan commented:
Rebased 15ff9cb1789b7fe69c0ad24d737712aa6955e9cb -> df0eb65bd98f300618cccd667720f1ccc6145865 ([spendblock_4](https://github.com/TheCharlatan/bitcoin/tree/spendblock_4) -> [spendblock_5](https://github.com/TheCharlatan/bitcoin/tree/spendblock_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_4..spendblock_5))

* Fixed conflict with #32644

TheCharlatan commented:
Rebased df0eb65bd98f300618cccd667720f1ccc6145865 -> 969799291d34da706b43209ba1209038ec349424 ([spendblock_5](https://github.com/TheCharlatan/bitcoin/tree/spendblock_5) -> [spendblock_6](https://github.com/TheCharlatan/bitcoin/tree/spendblock_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_5..spendblock_6))

* Fixed conflict with #32673
* Fixed conflict with #32421

TheCharlatan commented:
Rebased 969799291d34da706b43209ba1209038ec349424 -> 6e90a0eb605b5477b4e91128e08ed1cc5501a939 ([spendblock_6](https://github.com/TheCharlatan/bitcoin/tree/spendblock_6) -> [spendblock_7](https://github.com/TheCharlatan/bitcoin/tree/spendblock_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_6..spendblock_7))

* Fixed conflict with #31981

TheCharlatan commented:
Updated 1f96d1cefc721b579b7aa3b05c2e65df62c0c5cf -> 10c3a0689a3b73eb2b291ee81dd85066fa32497d ([spendblock_8](https://github.com/TheCharlatan/bitcoin/tree/spendblock_8) -> [spendblock_9](https://github.com/TheCharlatan/bitcoin/tree/spendblock_9), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_8..spendblock_9))

* Fixed silent merge conflict with #32521

ajtowns commented:
> Pass a block's spent coins through the existing `CBlockUndo` data structure to `ConnectBlock`

I don't think this is really a good direction. We already have a generic interface for supplying info about unspent coins: `CCoinsView`; I think we should be using that, rather than requiring users to convert their utxo data store into our undo format?

The thing that gets in the way of that seems to be the `AccessCoin` function returning a `Coin` by reference rather than value, but that doesn't seem particularly more efficient (it's an 8 byte copy rather than a 56 byte one unless the coin being spent is very non-standard, at least since #32279), and also seems a bit risky (in that the return value's lifetime is dependent on when the cache gets invalidated which is somewhat unpredictable).

But that seems to be fixable: https://github.com/ajtowns/bitcoin/commits/202509-ccoinsview/

(When AccessCoin was introduced in #1677, CScript always involved an additional heap allocation as it was just a `vector<char>`, so the reference would have saved an allocation. That changed a few years later when prevector was introduced in #6914 which allowed small scripts to not require an allocation. However, I haven't actually checked to see if the 8-vs-56-byte issue is actually as trivial as I think it should be)

TheCharlatan commented:
My hesitation was mostly about re-using the coins cache as is. In my opinion it is the wrong data structure to use in this case. Potential clients of the kernel are often capable of supplying the coins in order as consumed by a transaction's inputs. Requiring these to be stored in a map seems wrong. I.e. if a utreexo client has all the leaf data to go with a transaction, it should not have to use a less efficient data structure that consumes more memory to pass in the coins. But maybe we have to build out some of the usages here in order to see what the actual drawbacks are? My current thinking is though that just passing in the required coins contiguously, i.e. use the undo data structure, is the more pragmatic solution for most external users that do not wish to implement a utxo set. 

However this is also not exactly what I think you are suggesting here, since the `CCoinsView` could eventually provide a view over contiguous data too. As implemented at the moment, that seems more confusing to use though compared to a nested vector of coins.

Another alternative solution that was suggested by cfields is introducing a new transaction type that holds the coins alongside the prevouts. I think that would be an interesting alternative too.

purpleKarrot commented:
A function that takes a `const std::span<T> coins` argument cannot modify `coins`, but it has mutable access to `coins[x]`.

TheCharlatan commented:
Rebased 5988c7172406e3caca32fc78c011d878a1abdf1e -> 6493e47fd8d32c030d553f791caa48fcdb848c0b ([spendblock_14](https://github.com/TheCharlatan/bitcoin/tree/spendblock_14) -> [spendblock_15](https://github.com/TheCharlatan/bitcoin/tree/spendblock_15), [compare](https://github.com/TheCharlatan/bitcoin/compare/spendblock_14..spendblock_15))

ismaelsadeeq commented on src/consensus/tx_verify.cpp:
```diff
@@ -150,27 +150,33 @@ template unsigned int GetP2SHSigOpCount<const Coin>(
 template unsigned int GetP2SHSigOpCount<std::reference_wrapper<const Coin>>(
     const CTransaction& tx, const std::span<std::reference_wrapper<const Coin>>);
 
```
In "consensus: Use Coin span in GetTransactionSigOpCost" 3bdb4a89a9e852ddee9cebaaf3492473b16e7cc6

The last paragraph in the commit message is not stale from prev commit?

ismaelsadeeq commented on src/consensus/tx_verify.h:
```diff
@@ -56,12 +56,13 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const std::span<T> coins)
- * @param[in] tx     Transaction for which we are computing the cost
- * @param[in] inputs Map of previous transactions that have outputs we're spending
+ * @param[in] tx    Transaction for which we are computing the cost
+ * @param[in] coins Sorted span of Coins containing previous transaction outputs we're spending
```
In "consensus: Use Coin span in GetP2SHSigOpCount" 5709d07

Here and other commits.
Sorted in what order?

ismaelsadeeq reviewed (commented):
Concept ACK

ACK first four commits, I think the remaining commits can even be made better.
comments are inline

sedited commented on src/validation.cpp:
```diff
@@ -2899,6 +2809,127 @@ static void UpdateTipLog(
+
+    // verify that the view's current state corresponds to the previous block
+    uint256 hashPrevBlock = pindex->pprev == nullptr ? uint256() : pindex->pprev->GetBlockHash();
+    assert(hashPrevBlock == view.GetBestBlock());
```
I think I'd rather keep these duplicated for now. These are short-circuits to our logic, similar to how the coinbase checks work in various places. Based on that I'd rather keep them in place. This could also easily be improved in a follow-up.

ismaelsadeeq reviewed (approved):
Code review ACK a0e9578ec1ff1d8ec57f76a37e389aeeb554a639

sedited commented:
Rebased 00da3f04701789a6a238c3bee67e389396c01d6a -> 7802fd1f0d57bc20de0d71020c4c000baa65b217 ([spendblock_17](https://github.com/sedited/bitcoin/tree/spendblock_17) -> [spendblock_18](https://github.com/sedited/bitcoin/tree/spendblock_18), [compare](https://github.com/sedited/bitcoin/compare/spendblock_17..spendblock_18))

* Fixed conflict with #34164

sedited commented:
Updated 4f92e4e9d3d55ecb51795e9992aaf50be4feb4db -> 8140cba28afbeadf67b2db813f50267101c14b5e ([spendblock_19](https://github.com/sedited/bitcoin/tree/spendblock_19) -> [spendblock_20](https://github.com/sedited/bitcoin/tree/spendblock_20), [compare](https://github.com/sedited/bitcoin/compare/spendblock_19..spendblock_20))

* Addressed @w0xlt's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r2819423976), added a validation step checking that a block is not empty in the introduced `SpendBlock()` function.

sedited commented on src/validation.cpp:
```diff
@@ -2285,57 +2287,24 @@ script_verify_flags GetBlockScriptFlags(const CBlockIndex& block_index, const Ch
-
-    uint256 block_hash{block.GetHash()};
-    assert(*pindex->phashBlock == block_hash);
+    Assume(pindex->GetBlockHash() == block.GetHash());
```
I reverted it.

theuni reviewed (commented):
Concept ACK.

I agree that this is scary and requires a very comprehensive review (which I have not yet done here), but I think it's worth the effort. Being able to bring your own coins to the core validation functionality would make things much easier to reason about, and is (imo) how it should've been implemented all along.

sedited committed:
validation: Consolidate AccessCoins calls in PreChecks

This should introduce no behaviour change, but avoids a few extra vector
allocations and coins lookups

w0xlt commented on src/validation.cpp:
```diff
@@ -3033,7 +3095,17 @@ bool Chainstate::ConnectTip(
-        bool rv = ConnectBlock(*block_to_connect, state, pindexNew, view);
+        CBlockUndo blockundo;
+        if (!SpendBlock(*block_to_connect, pindexNew, view, state, blockundo)) {
+            assert(state.IsInvalid());
```
The current code assumes that `SpendBlock` can only fail when `state.IsInvalid()` is true.
However, in the `BLOCK_MUTATED` path, `SpendBlock` calls `FatalError()`, which sets the state to `IsError()` instead of `IsInvalid()`.
As a result, this assertion can cause the node to crash.

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index be2fb8687d..8e1d07b74b 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
+++ b/src/validation.cpp
@@ -3097,11 +3097,12 @@ bool Chainstate::ConnectTip(
         const auto reset_guard{view.CreateResetGuard()};
         CBlockUndo blockundo;
         if (!SpendBlock(*block_to_connect, pindexNew, view, state, blockundo)) {
-            assert(state.IsInvalid());
             if (m_chainman.m_options.signals) {
                 m_chainman.m_options.signals->BlockChecked(block_to_connect, state);
             }
-            InvalidBlockFound(pindexNew, state);
+            if (state.IsInvalid()) {
+                InvalidBlockFound(pindexNew, state);
+            }
             LogError("%s: SpendBlock %s failed, %s\n", __func__, pindexNew->GetBlockHash().ToString(), state.ToString());
             return false;
         }
```

Suggested test:

<details>
  <summary>diff</summary>

```diff
--- a/src/test/validation_chainstate_tests.cpp
+++ b/src/test/validation_chainstate_tests.cpp
@@ -4,8 +4,10 @@
 //
 #include <chainparams.h>
 #include <consensus/amount.h>
+#include <consensus/merkle.h>
 #include <consensus/validation.h>
 #include <node/kernel_notifications.h>
+#include <pow.h>
 #include <random.h>
 #include <rpc/blockchain.h>
 #include <script/script.h>
@@ -192,4 +194,39 @@ BOOST_FIXTURE_TEST_CASE(spendblock_rejects_empty_block, TestChain100Setup)
     BOOST_CHECK_EQUAL(state.GetRejectReason(), "high-hash");
 }
 
+BOOST_FIXTURE_TEST_CASE(activate_best_chain_does_not_invalidate_mutated_spendblock_failure, TestChain100Setup)
+{
+    Chainstate& chainstate = Assert(m_node.chainman)->ActiveChainstate();
+
+    auto block = std::make_shared<CBlock>(CreateBlock({}, CScript{} << OP_TRUE, chainstate));
+    block->hashMerkleRoot = uint256{1};
+    BOOST_REQUIRE(block->hashMerkleRoot != BlockMerkleRoot(*block));
+    block->nNonce = 0;
+    while (!CheckProofOfWork(block->GetHash(), block->nBits, Params().GetConsensus())) ++block->nNonce;
+
+    CBlockIndex index_dummy{*block};
+    const uint256 block_hash{block->GetHash()};
+
+    {
+        LOCK(cs_main);
+        CBlockIndex* tip = Assert(chainstate.m_chain.Tip());
+        index_dummy.pprev = tip;
+        index_dummy.nHeight = tip->nHeight + 1;
+        index_dummy.phashBlock = &block_hash;
+        index_dummy.nChainWork = tip->nChainWork + GetBlockProof(index_dummy);
+        index_dummy.nTx = block->vtx.size();
+        index_dummy.m_chain_tx_count = tip->m_chain_tx_count + index_dummy.nTx;
+        index_dummy.nStatus = BlockStatus(BLOCK_VALID_TRANSACTIONS | BLOCK_HAVE_DATA);
+        chainstate.setBlockIndexCandidates.insert(&index_dummy);
+    }
+
+    BlockValidationState state;
+    BOOST_CHECK(!chainstate.ActivateBestChain(state, block));
+    BOOST_CHECK(state.IsError());
+    BOOST_CHECK_EQUAL(state.GetResult(), BlockValidationResult::BLOCK_MUTATED);
+    BOOST_CHECK(!(index_dummy.nStatus & BLOCK_FAILED_VALID));
+
+    WITH_LOCK(cs_main, chainstate.setBlockIndexCandidates.erase(&index_dummy));
+}
+
```

</details>

sedited replied:
Thank you! That test seems like something we should add independently of this PR. Do you want to open a fresh PR for it? I can also do it, if you don't have the immediate capacity.

EDIT: It is also interesting that this is apparently the first test we have triggering this class of error. Or at least, it is currently the only test that prints an error string, which makes me suspect a suppression for it was not required by the other tests.

w0xlt replied:
Thanks! I can open a new PR for that test if needed, but since it’s already included here, I think it fits well with these changes.

sedited commented:
Thank you for the review and test case @w0xlt!

Rebased c361e1283c068b0e78837695d794ed3612499870 -> aa92c575ffbef12502f9d020379f5ecc39e0458c ([spendblock_22](https://github.com/sedited/bitcoin/tree/spendblock_22) -> [spendblock_23](https://github.com/sedited/bitcoin/tree/spendblock_23), [compare](https://github.com/sedited/bitcoin/compare/spendblock_22..spendblock_23))

Updated aa92c575ffbef12502f9d020379f5ecc39e0458c -> edc952b200285a197a2c0dbc4e57239514447155 ([spendblock_23](https://github.com/sedited/bitcoin/tree/spendblock_23) -> [spendblock_24](https://github.com/sedited/bitcoin/tree/spendblock_24), [compare](https://github.com/sedited/bitcoin/compare/spendblock_23..spendblock_24))

* Addressed @w0xlt's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r3070009097), adding the test and fixing the error case on block mutation.

sedited commented:
Re https://github.com/bitcoin/bitcoin/pull/32317#issuecomment-4282414534

Thank you for your summary @ismaelsadeeq . For completeness, I also have a pull request implementing validation with this PR opened on my rust crate: https://github.com/sedited/rust-bitcoinkernel/pull/98. If there is demand for doing (i) in the interim, I'll gladly move forward with that by reworking that pull request.

> I am inclined to suggest that the refactoring work in goal (ii) proceed methodically once we have expanded the tests in https://github.com/bitcoin/bitcoin/pull/35000 to cover not just TestBlockValidity but test/simulate other validation paths, so that we can catch bugs early. Something like https://github.com/bitcoin/bitcoin/pull/34895 should similarly be expanded to cover block validity and other block validation paths beyond just ConnectBlock see https://github.com/ismaelsadeeq/bitcoin/tree/fixup-connect-block-harness that expanded it to test block validity and testblockvalidity with spent outputs.

Yes, we clearly need some tests covering the block mutations, and if we get more than that, that would be good too. I would like review to continue here in the meantime.

sedited commented:
Opened #35187, which takes parts of aj's suggestion for wrapping the coins in a `CCoinsViewCache` and my previous work for exposing this PR's `SpendBlock`.

sedited commented:
Rebased edc952b200285a197a2c0dbc4e57239514447155 -> f047668025f611d80d06cc9f6b042ec948f1e034 ([spendblock_24](https://github.com/sedited/bitcoin/tree/spendblock_24) -> [spendblock_25](https://github.com/sedited/bitcoin/tree/spendblock_25), [compare](https://github.com/sedited/bitcoin/compare/spendblock_24..spendblock_25))

sedited commented:
Rebased f047668025f611d80d06cc9f6b042ec948f1e034 -> 8217d100a29c2c64a93d7b1e9544aef73ce695f7 ([spendblock_25](https://github.com/sedited/bitcoin/tree/spendblock_25) -> [spendblock_26](https://github.com/sedited/bitcoin/tree/spendblock_26), [compare](https://github.com/sedited/bitcoin/compare/spendblock_25..spendblock_26))

w0xlt commented on src/test/validation_chainstate_tests.cpp:
```diff
@@ -173,4 +178,64 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
+    LOCK(cs_main);
+    CBlockIndex* tip = Assert(chainstate.m_chain.Tip());
+
+    CBlock block;
```
The `high-hash` error means the block failed proof-of-work in `CheckBlockHeader()` before `CheckBlock()` reached the empty-block size check.

So this test is not actually proving that `SpendBlock()` handles empty blocks safely. It only proves that `SpendBlock()` calls `CheckBlock()` and rejects a header with invalid PoW.

That matters because the original risk was an empty block reaching later `SpendBlock()` logic and causing `block.vtx.size() - 1` underflow. If the test fails earlier with `high-hash`, it would still pass even if the empty-block size check path were broken or unreachable.

Suggested change:

<details>
  <summary>diff</summary>
 
```diff
diff --git a/src/test/validation_chainstate_tests.cpp b/src/test/validation_chainstate_tests.cpp
index 8e8726ce98..9f69b057c6 100644
--- a/src/test/validation_chainstate_tests.cpp
+++ b/src/test/validation_chainstate_tests.cpp
@@ -219,9 +219,13 @@ BOOST_FIXTURE_TEST_CASE(spendblock_rejects_empty_block, TestChain100Setup)
     LOCK(cs_main);
     CBlockIndex* tip = Assert(chainstate.m_chain.Tip());
 
-    CBlock block;
-    // CheckBlock() rejects empty blocks, ensure it gets run in SpendBlock
-    block.hashPrevBlock = tip->GetBlockHash();
+    CBlock block{CreateBlock({}, CScript{} << OP_TRUE)};
+    // CheckBlock() rejects empty blocks; keep the header otherwise valid so
+    // SpendBlock reaches the size check instead of failing earlier on PoW.
+    block.vtx.clear();
+    block.hashMerkleRoot = BlockMerkleRoot(block);
+    block.nNonce = 0;
+    while (!CheckProofOfWork(block.GetHash(), block.nBits, Params().GetConsensus())) ++block.nNonce;
 
     CBlockIndex index_dummy{block};
     const uint256 block_hash{block.GetHash()};
@@ -235,7 +239,7 @@ BOOST_FIXTURE_TEST_CASE(spendblock_rejects_empty_block, TestChain100Setup)
 
     BOOST_CHECK(!chainstate.SpendBlock(block, &index_dummy, view_dummy, state, blockundo));
     BOOST_CHECK(state.IsInvalid());
-    BOOST_CHECK_EQUAL(state.GetRejectReason(), "high-hash");
+    BOOST_CHECK_EQUAL(state.GetRejectReason(), "bad-blk-length");
 }
 
 BOOST_AUTO_TEST_SUITE_END()
```
</details>

w0xlt reviewed (commented):
ACK 8217d100a29c2c64a93d7b1e9544aef73ce695f7

sedited commented:
Updated 8217d100a29c2c64a93d7b1e9544aef73ce695f7 -> b07f73205e871402da2ccdb6a0d4c173dc6bfb69 ([spendblock_26](https://github.com/sedited/bitcoin/tree/spendblock_26) -> [spendblock_27](https://github.com/sedited/bitcoin/tree/spendblock_27), [compare](https://github.com/sedited/bitcoin/compare/spendblock_26..spendblock_27))

* Addressed @w0xlt's [comment](https://github.com/bitcoin/bitcoin/pull/32317#discussion_r3329333154), applied the change adding a valid-pow, empty block to exercise the desired code path in the spendblock rejects empty block test.

w0xlt reviewed (commented):
reACK b07f73205e871402da2ccdb6a0d4c173dc6bfb69

I would suggest splitting two responsibilities inside `Chainstate::ConnectBlock`:

- `ValidateBlockUndo(...)`: a private helper that validates the block using the provided undo data.
- `ConnectBlock(...)`: calls `ValidateBlockUndo(...)`, then writes undo data and updates block-index state if validation succeeds.

<details>
<summary>diff</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 6f675a9740..6e7a760051 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2295,15 +2295,28 @@ script_verify_flags GetBlockScriptFlags(const CBlockIndex& block_index, const Ch
     return flags;
 }
 
-bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo, BlockValidationState& state,
-                              CBlockIndex* pindex, bool fJustCheck)
+struct Chainstate::ConnectBlockValidationResult {
+    uint256 block_hash;
+    int n_inputs{0};
+    int64_t n_sigops_cost{0};
+    SteadyClock::time_point time_start{};
+    SteadyClock::time_point time_validation_done{};
+    bool is_genesis{false};
+};
+
+bool Chainstate::ValidateBlockUndo(
+    const CBlock& block,
+    const CBlockUndo& blockundo,
+    BlockValidationState& state,
+    const CBlockIndex& block_index,
+    bool fJustCheck,
+    ConnectBlockValidationResult& result)
 {
     AssertLockHeld(cs_main);
-    assert(pindex);
-    assert(pindex->GetBlockHash() == block.GetHash());
-    auto block_hash{pindex->GetBlockHash()};
+    assert(block_index.GetBlockHash() == block.GetHash());
+    result.block_hash = block_index.GetBlockHash();
 
-    const auto time_start{SteadyClock::now()};
+    result.time_start = SteadyClock::now();
     const CChainParams& params{m_chainman.GetParams()};
 
     assert(block.vtx.size() == blockundo.vtxundo.size() + 1); // blockundo does not contain a CTxUndo for the coinbase
@@ -2312,7 +2325,8 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
 
     // Special case for the genesis block, skipping connection of its transactions
     // (its coinbase is unspendable)
-    if (block_hash == params.GetConsensus().hashGenesisBlock) {
+    if (result.block_hash == params.GetConsensus().hashGenesisBlock) {
+        result.is_genesis = true;
         return true;
     }
 
@@ -2329,13 +2343,13 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
         BlockMap::const_iterator it{m_blockman.m_block_index.find(m_chainman.AssumedValidBlock())};
         if (it == m_blockman.m_block_index.end()) {
             script_check_reason = "assumevalid hash not in headers";
-        } else if (it->second.GetAncestor(pindex->nHeight) != pindex) {
-            script_check_reason = (pindex->nHeight > it->second.nHeight) ? "block height above assumevalid height" : "block not in assumevalid chain";
-        } else if (m_chainman.m_best_header->GetAncestor(pindex->nHeight) != pindex) {
+        } else if (it->second.GetAncestor(block_index.nHeight) != &block_index) {
+            script_check_reason = (block_index.nHeight > it->second.nHeight) ? "block height above assumevalid height" : "block not in assumevalid chain";
+        } else if (m_chainman.m_best_header->GetAncestor(block_index.nHeight) != &block_index) {
             script_check_reason = "block not in best header chain";
         } else if (m_chainman.m_best_header->nChainWork < m_chainman.MinimumChainWork()) {
             script_check_reason = "best header chainwork below minimumchainwork";
-        } else if (GetBlockProofEquivalentTime(*m_chainman.m_best_header, *pindex, *m_chainman.m_best_header, params.GetConsensus()) <= TWO_WEEKS_IN_SECONDS) {
+        } else if (GetBlockProofEquivalentTime(*m_chainman.m_best_header, block_index, *m_chainman.m_best_header, params.GetConsensus()) <= TWO_WEEKS_IN_SECONDS) {
             script_check_reason = "block too recent relative to best header";
         } else {
             // This block is a member of the assumed verified chain and an ancestor of the best header.
@@ -2357,20 +2371,20 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
     }
 
     const auto time_1{SteadyClock::now()};
-    m_chainman.time_check += time_1 - time_start;
+    m_chainman.time_check += time_1 - result.time_start;
     LogDebug(BCLog::BENCH, "    - Sanity checks: %.2fms [%.2fs (%.2fms/blk)]\n",
-             Ticks<MillisecondsDouble>(time_1 - time_start),
+             Ticks<MillisecondsDouble>(time_1 - result.time_start),
              Ticks<SecondsDouble>(m_chainman.time_check),
              Ticks<MillisecondsDouble>(m_chainman.time_check) / m_chainman.num_blocks_total);
 
     // Enforce BIP68 (sequence locks)
     int nLockTimeFlags = 0;
-    if (DeploymentActiveAt(*pindex, m_chainman, Consensus::DEPLOYMENT_CSV)) {
+    if (DeploymentActiveAt(block_index, m_chainman, Consensus::DEPLOYMENT_CSV)) {
         nLockTimeFlags |= LOCKTIME_VERIFY_SEQUENCE;
     }
 
     // Get the script flags for this block
-    script_verify_flags flags{GetBlockScriptFlags(*pindex, m_chainman)};
+    script_verify_flags flags{GetBlockScriptFlags(block_index, m_chainman)};
 
     const auto time_2{SteadyClock::now()};
     m_chainman.time_forks += time_2 - time_1;
@@ -2384,10 +2398,10 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
     if (script_check_reason != m_last_script_check_reason_logged && role.validated && !role.historical) {
         if (fScriptChecks) {
             LogInfo("Enabling script verification at block #%d (%s): %s.",
-                    pindex->nHeight, block_hash.ToString(), script_check_reason);
+                    block_index.nHeight, result.block_hash.ToString(), script_check_reason);
         } else {
             LogInfo("Disabling script verification at block #%d (%s).",
-                    pindex->nHeight, block_hash.ToString());
+                    block_index.nHeight, result.block_hash.ToString());
         }
         m_last_script_check_reason_logged = script_check_reason;
     }
@@ -2403,13 +2417,11 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
 
     std::vector<int> prevheights;
     CAmount nFees = 0;
-    int nInputs = 0;
-    int64_t nSigOpsCost = 0;
     for (unsigned int i = 0; i < block.vtx.size(); i++)
     {
         if (!state.IsValid()) break;
         const CTransaction &tx = *(block.vtx[i]);
-        nInputs += tx.vin.size();
+        result.n_inputs += tx.vin.size();
 
         if (!tx.IsCoinBase())
         {
@@ -2417,7 +2429,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
             CAmount txfee = 0;
             TxValidationState tx_state;
 
-            if (!Consensus::CheckTxInputs(tx, tx_state, spent_coins, pindex->nHeight, txfee)) {
+            if (!Consensus::CheckTxInputs(tx, tx_state, spent_coins, block_index.nHeight, txfee)) {
                 // Any transaction validation failure in ConnectBlock is a block consensus failure
                 state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
                               tx_state.GetRejectReason(),
@@ -2439,7 +2451,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
                 prevheights[j] = spent_coins[j].nHeight;
             }
 
-            if (!SequenceLocks(tx, nLockTimeFlags, prevheights, *pindex)) {
+            if (!SequenceLocks(tx, nLockTimeFlags, prevheights, block_index)) {
                 state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-txns-nonfinal",
                               "contains a non-BIP68-final transaction " + tx.GetHash().ToString());
                 break;
@@ -2451,11 +2463,11 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
         // * p2sh (when P2SH enabled in flags and excludes coinbase)
         // * witness (when witness enabled in flags and excludes coinbase)
         if (!tx.IsCoinBase()) {
-            nSigOpsCost += GetTransactionSigOpCost(tx, std::span<const Coin>{blockundo.vtxundo[i - 1].vprevout}, flags);
+            result.n_sigops_cost += GetTransactionSigOpCost(tx, std::span<const Coin>{blockundo.vtxundo[i - 1].vprevout}, flags);
         } else {
-            nSigOpsCost += GetTransactionSigOpCost(tx, std::span<const Coin>{}, flags);
+            result.n_sigops_cost += GetTransactionSigOpCost(tx, std::span<const Coin>{}, flags);
         }
-        if (nSigOpsCost > MAX_BLOCK_SIGOPS_COST) {
+        if (result.n_sigops_cost > MAX_BLOCK_SIGOPS_COST) {
             state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-blk-sigops", "too many sigops");
             break;
         }
@@ -2493,11 +2505,11 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
     m_chainman.time_connect += time_3 - time_2;
     LogDebug(BCLog::BENCH, "      - Connect %u transactions: %.2fms (%.3fms/tx, %.3fms/txin) [%.2fs (%.2fms/blk)]\n", (unsigned)block.vtx.size(),
              Ticks<MillisecondsDouble>(time_3 - time_2), Ticks<MillisecondsDouble>(time_3 - time_2) / block.vtx.size(),
-             nInputs <= 1 ? 0 : Ticks<MillisecondsDouble>(time_3 - time_2) / (nInputs - 1),
+             result.n_inputs <= 1 ? 0 : Ticks<MillisecondsDouble>(time_3 - time_2) / (result.n_inputs - 1),
              Ticks<SecondsDouble>(m_chainman.time_connect),
              Ticks<MillisecondsDouble>(m_chainman.time_connect) / m_chainman.num_blocks_total);
 
-    CAmount blockReward = nFees + GetBlockSubsidy(pindex->nHeight, params.GetConsensus());
+    CAmount blockReward = nFees + GetBlockSubsidy(block_index.nHeight, params.GetConsensus());
     if (block.vtx[0]->GetValueOut() > blockReward && state.IsValid()) {
         state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-cb-amount",
                       strprintf("coinbase pays too much (actual=%d vs limit=%d)", block.vtx[0]->GetValueOut(), blockReward));
@@ -2514,12 +2526,32 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
     }
     const auto time_4{SteadyClock::now()};
     m_chainman.time_verify += time_4 - time_2;
-    LogDebug(BCLog::BENCH, "    - Verify %u txins: %.2fms (%.3fms/txin) [%.2fs (%.2fms/blk)]\n", nInputs - 1,
+    LogDebug(BCLog::BENCH, "    - Verify %u txins: %.2fms (%.3fms/txin) [%.2fs (%.2fms/blk)]\n", result.n_inputs - 1,
              Ticks<MillisecondsDouble>(time_4 - time_2),
-             nInputs <= 1 ? 0 : Ticks<MillisecondsDouble>(time_4 - time_2) / (nInputs - 1),
+             result.n_inputs <= 1 ? 0 : Ticks<MillisecondsDouble>(time_4 - time_2) / (result.n_inputs - 1),
              Ticks<SecondsDouble>(m_chainman.time_verify),
              Ticks<MillisecondsDouble>(m_chainman.time_verify) / m_chainman.num_blocks_total);
 
+    result.time_validation_done = time_4;
+
+    return true;
+}
+
+bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo, BlockValidationState& state,
+                              CBlockIndex* pindex, bool fJustCheck)
+{
+    AssertLockHeld(cs_main);
+    assert(pindex);
+
+    ConnectBlockValidationResult validation_result;
+    if (!ValidateBlockUndo(block, blockundo, state, *pindex, fJustCheck, validation_result)) {
+        return false;
+    }
+
+    if (validation_result.is_genesis) {
+        return true;
+    }
+
     if (fJustCheck) {
         return true;
     }
@@ -2529,9 +2561,9 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
     }
 
     const auto time_5{SteadyClock::now()};
-    m_chainman.time_undo += time_5 - time_4;
+    m_chainman.time_undo += time_5 - validation_result.time_validation_done;
     LogDebug(BCLog::BENCH, "    - Write undo data: %.2fms [%.2fs (%.2fms/blk)]\n",
-             Ticks<MillisecondsDouble>(time_5 - time_4),
+             Ticks<MillisecondsDouble>(time_5 - validation_result.time_validation_done),
              Ticks<SecondsDouble>(m_chainman.time_undo),
              Ticks<MillisecondsDouble>(m_chainman.time_undo) / m_chainman.num_blocks_total);
 
@@ -2548,12 +2580,12 @@ bool Chainstate::ConnectBlock(const CBlock& block, const CBlockUndo& blockundo,
              Ticks<MillisecondsDouble>(m_chainman.time_index) / m_chainman.num_blocks_total);
 
     TRACEPOINT(validation, block_connected,
-        block_hash.data(),
+        validation_result.block_hash.data(),
         pindex->nHeight,
         block.vtx.size(),
-        nInputs,
-        nSigOpsCost,
-        Ticks<std::chrono::nanoseconds>(time_5 - time_start)
+        validation_result.n_inputs,
+        validation_result.n_sigops_cost,
+        Ticks<std::chrono::nanoseconds>(time_5 - validation_result.time_start)
     );
 
     return true;
diff --git a/src/validation.h b/src/validation.h
index 3e69c8fe73..7785a84c3b 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -572,6 +572,15 @@ protected:
 
     std::optional<const char*> m_last_script_check_reason_logged GUARDED_BY(::cs_main){};
 
+    struct ConnectBlockValidationResult;
+    bool ValidateBlockUndo(
+        const CBlock& block,
+        const CBlockUndo& blockundo,
+        BlockValidationState& state,
+        const CBlockIndex& block_index,
+        bool fJustCheck,
+        ConnectBlockValidationResult& result) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
 public:
     //! Reference to a BlockManager instance which itself is shared across all
     //! Chainstate instances.
```
</details>

sedited commented:
> I would suggest splitting two responsibilities inside Chainstate::ConnectBlock:

Yes, I think this is a good change, but I'm not sure if it should be part of this pull request. Seems like this could be done as a purely mechanical follow-up?

sedited commented:
Rebased b07f73205e871402da2ccdb6a0d4c173dc6bfb69 -> ea94091bcbba33a47c04b678481f76b4a1608360 ([spendblock_27](https://github.com/sedited/bitcoin/tree/spendblock_27) -> [spendblock_28](https://github.com/sedited/bitcoin/tree/spendblock_28), [compare](https://github.com/sedited/bitcoin/compare/spendblock_27..spendblock_28))

* Fixed conflict with #35414

yuvicc commented on src/coins.h:
```diff
@@ -458,6 +458,21 @@ class CCoinsViewCache : public CCoinsViewBacked
+        for (const CTxIn& input : tx.vin) {
+            coins.emplace_back(AccessCoin(input.prevout));
+        }
+        return callback(std::span{coins});
```
minor nit: acc to [core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter), it would be good if we use `std::forward` here?

```suggestion
        return std::forward<Callable>(callback)(std::span{coins});
```

yuvicc reviewed (commented):
Approach ACK ea94091bcbba33a47c04b678481f76b4a1608360

Splitting utxo mutation into `SpendBlock` and leaving `ConnectBlock` as pure verification over is a nice separation.

sedited commented:
Thank you for the reviews and discussion (especially @w0xlt and @ajtowns), but I'm closing this again. Post parallel input fetcher I don't think there is a nice way to achieve the goals of this PR without materially changing its implementation. At that point I would rather open a new pull request. The overarching goal I want to achieve here can also be solved with #35187, which I would like people to focus their review on. @w0xlt do you want to open a separate PR with the unit tests introduced here?
